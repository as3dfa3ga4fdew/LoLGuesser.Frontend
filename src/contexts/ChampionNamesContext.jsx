import { createContext, useEffect, useState} from "react";
const ChampionNamesContext = createContext();

const ChampionNamesProvider = (props) =>{
    const championNamesUpdate = (val) =>{
        setChampionNames(val);
    }

    const [championNames, setChampionNames] = useState([]);

    useEffect(() => {
        GetChampionNames(setChampionNames);
    }, []);

    return(
        <ChampionNamesContext.Provider value={{championNames, championNamesUpdate}}>
            {props.children}
        </ChampionNamesContext.Provider>
    );
}

const GetChampionNames = async (setChampionNames) => {
    try
    {
        let response = await fetch("https://localhost:5000/api/Game/names");

        if(response.ok !== true)
        {
            //setErrorMessage("Ooops something went wrong please try again later... 1");
            return;
        }
        let result = await response.json();

        setChampionNames(JSON.parse(JSON.stringify(result)));
    }
    catch(error)
    {
        console.log(error);
        //setErrorMessage("Ooops something went wrong please try again later... 1");
    }
}



export { ChampionNamesProvider, ChampionNamesContext };