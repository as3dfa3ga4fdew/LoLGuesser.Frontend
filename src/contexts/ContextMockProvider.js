import React, { createContext, useContext } from 'react';

//Mock context for UserContext
export const UserContext = createContext();

export const UserContextMockProvider = ({ children, value }) => {
  return (
    <UserContext.Provider value={value}>
      {children}
    </UserContext.Provider>
  );
};

//Mock context for ChampionNamesContext
export const ChampionNamesContext = createContext();

export const ChampionNamesContextMockProvider = ({ children, value }) => {
  return (
    <ChampionNamesContext.Provider value={value}>
      {children}
    </ChampionNamesContext.Provider>
  );
};
