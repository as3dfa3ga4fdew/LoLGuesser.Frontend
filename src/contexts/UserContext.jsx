import { createContext, useContext, useState, useEffect } from "react"
import config from "../config.json";

const UserContext = createContext();

const UserProvider = (props) => {

  const keyName = "user";
  const userBuilder = {jwt: "", username: "", score: 0, rememberMe: false};

  const userUpdate = (val) => {
    setUser(val);
  }

  const [user, setUser] = useState(null);

  useEffect(() => {
    let value = JSON.parse(sessionStorage.getItem(keyName));

    if(value){
      setUser(value);
      return;
    }

    value = JSON.parse(localStorage.getItem(keyName));

    if(!value){
      setUser(userBuilder);
      return;
    }

    SetValidatedUserAsync(value.jwt, setUser, value, userBuilder);

  }, []);

  useEffect(() => {
    if(user === null)
      return;

    sessionStorage.setItem(keyName, JSON.stringify(user));

    if(user.rememberMe === false){
      localStorage.removeItem(keyName);
      return;
    }

    localStorage.setItem(keyName, JSON.stringify(user));

  }, [user]);

  return(
    <UserContext.Provider value={{user, userUpdate}}>
      {props.children}
    </UserContext.Provider>
  );
}

const SetValidatedUserAsync = async (jwt, setUser, value, userBuilder) => {
  try{
    let response = await fetch(config.serverUrl + "/api/auth/token", {
      headers: {
        Authorization: "Bearer " + jwt
      }
    });

    if(response.ok !== true){
      setUser(userBuilder);
      return;
    }
  }
  catch(error){
    setUser(userBuilder);
    console.log(error);
  }

  setUser(value);
  return;
}


export { UserProvider, UserContext}