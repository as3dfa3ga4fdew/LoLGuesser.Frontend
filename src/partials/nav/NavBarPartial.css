.nav-header{
    width: 100%;
    height: 80px;
    background-color: rgb(10, 20, 40);
    display: flex;
    align-items: center;
    padding-left: 15px;
    position: fixed;
}

.header-item{
    height: 50px;
    display: flex;
    align-items: center;
    justify-content: center;
    color: var(--gold-text-color);
    font-size: large;
    font-family: Arial, sans-serif;
}

.header-logo{
    height: 60px;
    width: 60px;
}

.header-account-items{
    display: flex;
    margin-left: 40px;
}

#header-item-right{
    margin-left: 40px;
}

.logout-button{
    height: 50px;
    display: flex;
    align-items: center;
    justify-content: center;
    color: var(--gold-text-color);
    background: none;
    border: none;
    text-decoration: underline;
    font-size: large;
    font-weight: lighter;
    font-family: Arial, sans-serif;
    padding-right: 30px;
  }
  
  .logout-button:hover{
    cursor: pointer;
  }

  .user-info{
    display: flex;
    align-items: center;
    justify-content: center;
    margin-left: 10px;
    color: var(--gold-text-color);
  }