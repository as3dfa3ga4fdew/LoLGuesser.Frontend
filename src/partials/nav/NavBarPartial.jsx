import "./NavBarPartial.css";
import '../../globals/Global.css';
import { NavLink } from "react-router-dom";
import Logo from '../../images/logo.png'
import {UserContext} from "../../contexts/UserContext";
import { useContext } from "react";
const NavBarPartial = () =>{
    const {userUpdate, user} = useContext(UserContext);

    const logout = () =>{
        userUpdate({jwt: "", username: "", score: 0, rememberMe: false});
    }

    return(
        <nav className="nav-header">
            <NavLink to="/" >
                <img src={Logo} alt="" className="header-logo"/>
            </NavLink>
            {
                user == null || user.jwt == "" ? 
                <div className="header-account-items">
                    <NavLink className="header-item" to="/login" >Login</NavLink>
                    <NavLink className="header-item" id="header-item-right" to="/register" >Register</NavLink>
                </div>
                :
                <div className="header-account-items">
                    <button className="logout-button" onClick={logout}>Logout</button>
                    <div className="user-info">{user.username}:</div>
                    <div className="user-info">{user.score}</div>
                </div>
            }
          
        </nav>
    );
}

export default NavBarPartial;