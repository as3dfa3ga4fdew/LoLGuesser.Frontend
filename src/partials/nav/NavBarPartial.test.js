import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { UserContext } from "../../contexts/UserContext";
import NavBarPartial from './NavBarPartial';

describe('NavBarPartial', () => {
  const userUpdateMock = jest.fn();

  it('renders correctly when user is not logged in', () => {
    render(
      <UserContext.Provider value={{ user: null, userUpdate: userUpdateMock }}>
        <Router>
          <NavBarPartial />
        </Router>
      </UserContext.Provider>
    );

    // Expect the login and register links to be present
    expect(screen.getByText(/login/i)).toBeInTheDocument();
    expect(screen.getByText(/register/i)).toBeInTheDocument();
    // Expect Logout button to not be present
    expect(screen.queryByText(/logout/i)).not.toBeInTheDocument();
  });

  it('calls the logout function when the logout button is clicked', () => {
    const user = { jwt: "token", username: "testUser", score: 5, rememberMe: true };

    render(
      <UserContext.Provider value={{ user: user, userUpdate: userUpdateMock }}>
        <Router>
          <NavBarPartial />
        </Router>
      </UserContext.Provider>
    );

    fireEvent.click(screen.getByText(/logout/i));

    // Expect the userUpdate function to have been called with the initial state
    expect(userUpdateMock).toHaveBeenCalledWith({jwt: "", username: "", score: 0, rememberMe: false});
  });
});