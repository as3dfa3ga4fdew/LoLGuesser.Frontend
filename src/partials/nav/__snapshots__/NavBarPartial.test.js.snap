// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarPartial NavBarPartial snapshot test 1`] = `
<DocumentFragment>
  <nav
    class="nav-header"
  >
    <a
      aria-current="page"
      class="active"
      href="/"
    >
      <img
        alt="company logo"
        class="header-logo"
        src="logo.png"
      />
    </a>
    <div
      class="header-account-items"
    >
      <a
        class="header-item"
        href="/login"
      >
        Login
      </a>
      <a
        class="header-item"
        href="/register"
        id="header-item-right"
      >
        Register
      </a>
    </div>
  </nav>
</DocumentFragment>
`;
